tracepoint_decode/src/display.rs

```diff
@@ -0,0 +1,572 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+//! `fmt::Display` helpers for using  tracepoint_decode types with format
+//! macros like [`write!`] and [`format_args!`].
+
+use core::fmt;
+use core::fmt::Write;
+
+use eventheader_types::*;
+
+use crate::charconv;
+use crate::enumerator;
+use crate::filters;
+use crate::filters::Filter;
+use crate::perf_abi;
+use crate::perf_event_desc;
+use crate::perf_session;
+use crate::writers;
+
+use crate::PerfConvertOptions;
+use crate::PerfMetaOptions;
+
+/// Display implementation that JSON-escapes the provided input string.
+/// This escapes control chars, quotes, and backslashes. For example,
+/// the string `Hello, "world"!` would be displayed as `Hello, \"world\"!`.
+pub struct JsonEscapeDisplay<'str> {
+    value: &'str str,
+}
+
+impl<'str> JsonEscapeDisplay<'str> {
+    /// Creates a new formatter for the specified string.
+    pub fn new(value: &'str str) -> Self {
+        return Self { value };
+    }
+
+    /// Writes the JSON-escaped value to the specified writer.
+    pub fn write_to<W: fmt::Write + ?Sized>(&self, writer: &mut W) -> fmt::Result {
+        let mut dest = filters::WriteFilter::new(writer);
+        return filters::JsonEscapeFilter::new(&mut dest).write_str(self.value);
+    }
+}
+
+impl<'str> fmt::Display for JsonEscapeDisplay<'str> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> fmt::Result {
+        return self.write_to(f);
+    }
+}
+
+/// Display implementation for string data is expected to be UTF-8. For example,
+/// this may be used for the name of an EventHeader event or field.
+///
+/// Tries to interpret the value as UTF-8, but falls back to Latin1 if the value
+/// contains non-UTF-8 sequences. This allows the value to be displayed with
+/// best-effort fidelity even if the event is incorrectly-authored or corrupt.
+///
+/// Instances of this type are returned by methods such as
+/// [`crate::EventHeaderEventInfo::name_display`] and
+/// [`crate::EventHeaderItemInfo::name_display`].
+#[derive(Clone, Copy, Debug)]
+pub struct Utf8WithLatin1FallbackDisplay<'dat> {
```

Comment:
Note: Renamed this type from `NameDisplay` since it is potentially useful in situations other than just names. Added a public constructor so caller can use it themselves instead of just for names.